### PR TITLE
Implicitly convert XML rval to string

### LINKED ISSUE #2459: XML nodes are too picky about setting values
When setting an XML node content, PowerShell doesn't correctly unwrap PSObject strings.

For instance: `Join-Path` obviously outputs strings, but it outputs them wrapped in PSObjects
## Steps to reproduce

``` posh
[xml]$x  = "<root><path/></root>"
$x.root.path = Join-Path C:\ Windows
```
## Expected behavior

`$x.root.path` is an XML Node with string content "C:\Windows"
## Actual behavior

A `SetValueException` is thrown:

```
Cannot set "path" because only strings can be used as values to set XmlNode properties.
At line:2 char:1
+ $x.root.path = Join-Path C:\ Windows
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], SetValueException
    + FullyQualifiedErrorId : XmlNodeSetShouldBeAString
```
## Environment data

``` powershell
> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      5.1.14393.206
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.14393.206
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1

```
## Workaround

Either of the following would work, but should not be necessary:

``` posh
$x.root.path = (Join-Path C:\ Windows).PSObject.BaseObject
$x.root.path = (Join-Path C:\ Windows).ToString()
```

### PR DESCRIPTION
**Pull request by iSazonov**
1. Implicitly convert XML rval to string
2. Add test

Close #2459

### CODE REVIEW

File: src/System.Management.Automation/engine/CoreAdapter.cs [hunk 1]
```diff
@@ -4911,7 +4911,8 @@ protected override object PropertyGet(PSProperty property)
         /// <param name="convertIfPossible">instructs the adapter to convert before setting, if the adapter supports conversion</param>
         protected override void PropertySet(PSProperty property, object setValue, bool convertIfPossible)
         {
-            string valueString = setValue as string;
+            // XML is always a string so implicitly convert to string
+            string valueString = setValue?.ToString();
```
Comment (lzybkr) [inline src/System.Management.Automation/engine/CoreAdapter.cs]: Calling ToString directly avoids any magic PowerShell might apply when converting a value to string.

You could just call `LanguagePrimitives.ConvertTo<string>(setValue)` - this will properly call into a `CodeMethod` or `ScriptMethod` named `ToString` if the object has one.

File: test/powershell/engine/ETS/Adapter.Tests.ps1 [hunk 2]
```diff
@@ -80,3 +80,13 @@ Describe "Adapter Tests" -tags "CI" {
...
 }
+
+Describe "Adapter XML Tests" -tags "CI" {
+    It "Can set XML node property to non-string object" {
+        [xml]$x  = "<root><data/></root>"
+        $rval = @{c=1}
+        # rval will be implicitly converted to `string` type
+        { $x.root.data = $rval } | Should Not Throw
+        $x.root.data | Should Be $rval.ToString()
+    }
```
Comment (lzybkr) [inline test/powershell/engine/ETS/Adapter.Tests.ps1]: You should also test that we properly invoke `ToString` implemented as a `CodeMethod` and `ScriptMethod`. `System.Type` has a `CodeMethod` implementation of `ToString`, you can see the difference:
```
#57 PS> [System.Int32].ToString()
System.Int32
#58 PS> [string][System.Int32]
int
```

**Review by lzybkr — CHANGES_REQUESTED** [review]

File: src/System.Management.Automation/engine/CoreAdapter.cs (hunk 1 above)
Comment (iSazonov) [inline src/System.Management.Automation/engine/CoreAdapter.cs]: Done.

File: test/powershell/engine/ETS/Adapter.Tests.ps1 (hunk 2 above)
Comment (iSazonov) [inline test/powershell/engine/ETS/Adapter.Tests.ps1]: Done.
Only I could not create `ToString`  ScriptMethod  because `ToString` already exists.

**Comment by iSazonov** [comment]
@lzybkr could continue code review?

File: src/System.Management.Automation/engine/CoreAdapter.cs [hunk 3]
```diff
@@ -4911,7 +4911,8 @@ protected override object PropertyGet(PSProperty property)
         /// <param name="convertIfPossible">instructs the adapter to convert before setting, if the adapter supports conversion</param>
         protected override void PropertySet(PSProperty property, object setValue, bool convertIfPossible)
         {
-            string valueString = setValue as string;
+            // XML is always a string so implicitly convert to string
+            string valueString = LanguagePrimitives.ConvertTo<string>(setValue);
             if (valueString == null)
```
Comment (lzybkr) [inline src/System.Management.Automation/engine/CoreAdapter.cs]: I think this condition is unreachable now.

We can convert everything to string (including `$null` which turns into the empty string).

The only failure condition now is if the conversion to string fails with an exception, e.g. if `ToString()` throws an exception. This should be rare, but if it happens, I think we'd want to see that exception and not the `SetValueException` that is being thrown here.

In other words, I think this test and subsequent throw can be safely deleted.

File: test/powershell/engine/ETS/Adapter.Tests.ps1 [hunk 4]
```diff
@@ -80,3 +80,24 @@ Describe "Adapter Tests" -tags "CI" {
...
 }
+
+Describe "Adapter XML Tests" -tags "CI" {
+    BeforeAll {
+        [xml]$x  = "<root><data/></root>"
+        $testCases =
+            @{ rval = @{testprop = 1}; value = 'hash (psobject)' },
+            @{ rval = 1;               value = 'int (codemethod)' },
+            @{ rval = "teststring";    value = 'string (codemethod)' }
+    }
```
Comment (lzybkr) [inline test/powershell/engine/ETS/Adapter.Tests.ps1]: I would add tests for:
* $null
* array of strings
* array of integers
* psobject wrapping each of the previous values

File: test/powershell/engine/ETS/Adapter.Tests.ps1 [hunk 5]
```diff
@@ -80,3 +80,24 @@ Describe "Adapter Tests" -tags "CI" {
...
+
+    Context "Can set XML node property to non-string object" {
+        It "rval is <value>" -TestCases $testCases {
+            # rval will be implicitly converted to 'string' type
+            param($rval)
+            {
+                { $x.root.data = $rval } | Should Not Throw
+                $x.root.data | Should Be $rval.ToString()
```
Comment (lzybkr) [inline test/powershell/engine/ETS/Adapter.Tests.ps1]: We don't always call `ToString`.

```powershell
PS> $x = 1,2
PS> [System.Management.Automation.LanguagePrimitives]::ConvertTo($x, [string])
1 2
PS> $x.ToString()
System.Object[]
```

File: test/powershell/engine/ETS/Adapter.Tests.ps1 (hunk 4 above)
Comment (iSazonov) [inline test/powershell/engine/ETS/Adapter.Tests.ps1]: I don't understand last `psobject wrapping each of the previous values`

File: test/powershell/engine/ETS/Adapter.Tests.ps1 (hunk 5 above)
Comment (iSazonov) [inline test/powershell/engine/ETS/Adapter.Tests.ps1]: Fixed.

**Comment by mirichmo** [comment]
@lzybkr Do you have any other comments or concerns regarding this PR?

**Review by lzybkr — APPROVED** [review]

**Comment by iSazonov** [comment]
@lzybkr Thanks for code review and great comments!
